Q0Q2 optimization

martin-ueding opened the pull request:
This resolves #96. It currently is a work in progress, this pull request starts for everyone wanting to track the progress.

martin-ueding committed:
Add scripts that went missing

martin-ueding committed:
Add more notes

martin-ueding committed:
Start with DilutedProductFactory

martin-ueding committed:
Add an untested DilutedProductFactoryQ0Q2

martin-ueding committed:
Instantiate DilutedProductFactory

martin-ueding committed:
Use DilutedProductFactory in DilutedTraceFactory6, crashes

martin-ueding committed:
Some more stuff

martin-ueding committed:
Merge branch 'master' into Q0Q2-optimization

martin-ueding committed:
Produce less timing output

martin-ueding committed:
Pass the Q0Q2 factory to the C4 traces as well

martin-ueding committed:
Add a summary to the comparison script

martin-ueding committed:
Port the C4 trace

martin-ueding committed:
Remove unused code

martin-ueding committed:
Go a step back to working C6cC

martin-ueding committed:
Use cached quantities for part of the C6cC

martin-ueding committed:
Rename variables

martin-ueding committed:
Merge branch 'master' into Q0Q2-optimization

martin-ueding committed:
Add operator* associativity test

martin-ueding committed:
Do not manually do powers ot two

martin-ueding committed:
Do comparison of numbers with NEAR only

martin-ueding committed:
More test output

martin-ueding committed:
Fix the associativity unit test

martin-ueding committed:
Just some debugging additions

martin-ueding committed:
Some debug code

martin-ueding committed:
Try fastest variant now

martin-ueding committed:
Actually use new operator* checks

martin-ueding committed:
Remove alternative versions

martin-ueding committed:
Only print on errors

martin-ueding committed:
Update C6cC reference data

I compared the three pion contractions at all momenta zero with
contractions that Liuming has performed a while ago. All diagrams match
except for the C6cC. Additionally we see that only that diagram suffered
from the strange `operator*` issues. Therefore we conclude that it was
wrong before and it is now fixed, therefore the reference data for that
diagram is updated now.

martin-ueding committed:
Remove now unused variables

martin-ueding committed:
Try compact Bash array append for Travis CI

martin-ueding committed:
Try to be more careful with variables

martin-ueding committed:
Also clear Q0Q2 cache

kostrzewa commented:
@martin-ueding @matfischer @pittlerf 

So this is basically ready to use and proven to produce correct results for all 2pt, 3pt and 4pt diagrams, correct?

Does it also reduce memory load by clearing the Q2 once all Q0Q2 have been constructed?

martin-ueding committed:
Reverse time in C6cC trace

martin-ueding committed:
Reverse time in reference data

martin-ueding committed:
Add reversal script

martin-ueding committed:
Change the C-LIME path

martin-ueding committed:
Expose all cutoffs in the config

martin-ueding committed:
Merge branch 'master' into Q0Q2-optimization

martin-ueding committed:
Merge branch 'Q0Q2-optimization' of github.com:HISKP-LQCD/sLapH-contractions into Q0Q2-optimization

martin-ueding committed:
Print out the number of combinations

martin-ueding committed:
Include missing header

martin-ueding committed:
Use member function

martin-ueding committed:
Merge branch 'master' into Q0Q2-optimization

martin-ueding committed:
Add JUWELS module script

martin-ueding committed:
Remove redundant build scripts

martin-ueding committed:
Strip away a bunch of unused options

martin-ueding committed:
Add utility script

martin-ueding committed:
Extract names from reference correlators

martin-ueding committed:
Use new correlator list in first unit test

martin-ueding committed:
Start with JSON parser

martin-ueding committed:
Add some notes

kostrzewa commented on Notes/GlobalData.md:
```diff
@@ -0,0 +1,264 @@
+
+---
+
+You know what's fun? Code archaeology! Let's go on a fun trip to code that has
```
at least it was only written by a few people :)

martin-ueding committed:
More notes

martin-ueding committed:
Fix JSON syntax

martin-ueding committed:
Just make the code compile but break results

martin-ueding committed:
Indent

martin-ueding committed:
Try to parse strings, fails

martin-ueding committed:
Fix parsing for two-digit gamma structures

martin-ueding committed:
Get it to work!!

martin-ueding committed:
Modify other unit tests

martin-ueding committed:
Update desired correlators in new test as it uses 6 up

martin-ueding committed:
Install Boost regex on Travis CI

martin-ueding committed:
Always use negative momenta VdaggerV

martin-ueding committed:
Improve condition

martin-ueding committed:
Fix stupid errors, thanks to group seminar

martin-ueding committed:
Use atomics to save memory

martin-ueding committed:
Take care of begneign warnings

martin-ueding committed:
Clear the Q2Q0 caches after each trace

martin-ueding committed:
Mark more stuff unused

martin-ueding committed:
Format

martin-ueding committed:
More formatting

martin-ueding committed:
Fix all warnings

martin-ueding committed:
Remove CMake warning I don't fully understand

martin-ueding committed:
Remove parallelization from the main loop

martin-ueding committed:
Change output message

martin-ueding committed:
Make DilutedFactorFactory::operator[] synchonized

martin-ueding committed:
Make the DilutedTraceFactory thread safe

martin-ueding committed:
Make the DilutedProductFactoryQ0Q2 thread safe

martin-ueding committed:
Parallelize assemble_impl

martin-ueding committed:
Parallelize DilutedTraceFactory

martin-ueding committed:
No need to have a lock a level higher now

martin-ueding committed:
Start preparing for eager building

martin-ueding committed:
Refactor the resolve_request

martin-ueding committed:
Try to use a set

martin-ueding committed:
Remove the parallelization for now

martin-ueding committed:
Get 3pi diagrams to work again

martin-ueding committed:
git diff

martin-ueding committed:
Let the DilutedProductFactory take requests

martin-ueding committed:
Start with multi layer scheduling

martin-ueding committed:
Introduce a variadic template for the number of times

martin-ueding committed:
Throw exceptions instead

martin-ueding committed:
Try other recursive template

martin-ueding committed:
Towards request_impl for every trace

martin-ueding committed:
Get it back working for charged diagrams

martin-ueding committed:
Reserve later

martin-ueding committed:
Pull out building to the top level

martin-ueding committed:
Delegate building in all members of DiagramParts

martin-ueding committed:
Remove delegation of building from lower levels

martin-ueding committed:
Add some more TimingScope

martin-ueding committed:
Add more timing scopes

martin-ueding committed:
Add parallelization to 6pt-trace factory

martin-ueding committed:
Add parallelization to all traces

martin-ueding committed:
Try to add more barriers, does not help

martin-ueding committed:
Merge branch 'inplace-timings' into Q0Q2-optimization

martin-ueding committed:
Merge branch 'Q0Q2-optimization' into other-parallelization

martin-ueding committed:
Include the source file again

martin-ueding committed:
Fix assert

martin-ueding committed:
Remove cache clearing option that does not have any effect now

martin-ueding committed:
Try the other GraphViz engines

martin-ueding committed:
Move some functions to the source files

martin-ueding committed:
Remove qualifier

martin-ueding committed:
Clear requests earlier

martin-ueding committed:
Add debug output

martin-ueding committed:
Multiply regardless of existence

The DilutedFactorsMap now has empty entries because they are
instantiated before. This means that they were skipped before and lead
to empty DilutedFactorMaps which lead to crashes in other parts of the
code. With the scheduling scheme we have there is no need to do it lazy
any more, we can just assume that they did not exist in the first place
and do the computations.

martin-ueding committed:
Mark nodes and edges in parallel sections in red

martin-ueding committed:
Also color edges, add vertical mode

martin-ueding committed:
Move git info to a source file for faster incremental building

martin-ueding committed:
Ignore popping of last dtor

martin-ueding committed:
Exclude C3c diagram in test for now

martin-ueding committed:
Comment out memory estimation

This does not work anyway, and I kept getting assertion failues there:

        trQ1Q1: 0.00 Gb
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 28) >= this->size() (which is 28)

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50        return ret;
\#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
\#1  0x00007ffff6a108d9 in __GI_abort () at abort.c:79
\#2  0x00007ffff6db26b6 in __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
\#3  0x00007ffff6dbe86c in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:47
\#4  0x00007ffff6dbe8c7 in std::terminate () at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:57
\#5  0x00007ffff6dbeb29 in __cxxabiv1::__cxa_throw (obj=<optimized out>, tinfo=0x7ffff6f00328 <typeinfo for std::out_of_range>,
    dest=0x7ffff6dd4ac0 <std::out_of_range::~out_of_range()>) at ../../../../libstdc++-v3/libsupc++/eh_throw.cc:95
\#6  0x00007ffff6db5001 in std::__throw_out_of_range_fmt (__fmt=<optimized out>)
    at ../../../../../libstdc++-v3/src/c++11/functexcept.cc:82
\#7  0x0000000000929455 in std::vector<DilutedFactorIndex, std::allocator<DilutedFactorIndex> >::_M_range_check (
    this=0x60800000e760, __n=28) at /usr/include/c++/9/bits/stl_vector.h:1070
\#8  0x0000000000924c73 in std::vector<DilutedFactorIndex, std::allocator<DilutedFactorIndex> >::at (this=0x60800000e760, __n=28)
    at /usr/include/c++/9/bits/stl_vector.h:1091
\#9  0x000000000091c299 in read_parameters (gd=..., ac=7, av=0x7fffffffe438)
    at /home/mu/Lattice/Code/sLapH-contractions/src/global_data.cpp:367
\#10 0x000000000083bb5e in main (ac=7, av=0x7fffffffe438) at /home/mu/Lattice/Code/sLapH-contractions/main/contract.cpp:41

martin-ueding committed:
Delete old XML timing stuff

martin-ueding committed:
Move some stuff into the utility directory

martin-ueding committed:
Move introduction

martin-ueding committed:
Merge Notes and doc directories

martin-ueding committed:
Ignore all other files

martin-ueding committed:
Parallelize DilutedFactorFactory

martin-ueding committed:
Remove quadratic waste of resources

martin-ueding committed:
Make operator< also work for different sizes

martin-ueding committed:
Add divisor and remainder

martin-ueding committed:
Use source slice and not time difference

martin-ueding committed:
Adjust JUWELS job script

martin-ueding committed:
Fix start_runs.sh script

martin-ueding committed:
Explicitly request memory

martin-ueding committed:
Merge branch 'other-parallelization' into Q0Q2-optimization

martin-ueding commented:
We have the problem that I have not touched the diagrams that I do not use. This means that the integration tests fail for all the neutral diagrams and the three point diagrams. The code therefore is not suitable for the rho, and we cannot merge this into `master` yet. But as I do not need it, I don't want to spent time working on that just yet. Perhaps we will just have to leave this pull request hanging until somebody actually needs the neutral diagrams.

martin-ueding committed:
Add test program for the strides

martin-ueding committed:
Remove NumPy dependency from job script generator

martin-ueding committed:
Add support for accounts

martin-ueding committed:
Fix parens error

martin-ueding committed:
Lowercase Jinja identifier

martin-ueding committed:
Make for loops compatible with older OpenMP

martin-ueding committed:
Change my email address

martin-ueding committed:
Merge branch 'Q0Q2-optimization' of github.com:HISKP-LQCD/sLapH-contractions into Q0Q2-optimization

martin-ueding committed:
Increase the job time to 4 days

martin-ueding committed:
Merge branch 'Q0Q2-optimization' of github.com:HISKP-LQCD/sLapH-contractions into Q0Q2-optimization

martin-ueding committed:
Change position of libraries

martin-ueding committed:
Try a middle_sum

martin-ueding committed:
Try text output

martin-ueding committed:
Set the output modifier more sensibly

martin-ueding committed:
Use correct setprecision instead

martin-ueding committed:
Actually check all mutual combinations of used random vectors

martin-ueding committed:
Update the correlator_names.js to include everyting from reference

martin-ueding committed:
Merge branch 'Q0Q2-optimization' of github.com:HISKP-LQCD/sLapH-contractions into Q0Q2-optimization

martin-ueding committed:
Use new list with correlators for unit test

kostrzewa committed:
instantiate right accumulator only after checking random vector combinations
